### ISSUE TITLE
Service not restarting when enabled = True on Ubuntu

### ISSUE THREAD

**Issue by vladvasiliu** [saltstack/salt#18311]
This is an issue for the problem reported on the mailing list: https://groups.google.com/forum/#!topic/salt-users/9cNJKiS2lQw

I am trying to manage Shorewall with Salt on Ubuntu 14.10, and when I set `enable: true` and `watch: [something]` in the service definition, the service is not restarted. If I remove `enable: true` then the service is restarted. This has happened with both versions 2014.1.10 and 2014.7.0.

Upon further investigation, I have found that this does not happen for other services, e.g if the service is named "nginx", everything works as expected. On CentOS 7, everything works as expected.

In the minion logs, it appears that `service.mod_watch` is not called when I have `enabled: true` for shorewall, but it is called for other service names and on CentOS.

Here are some dummy states allowing to reproduce the issue:

`nginx.sls`
```
firewall:
    pkg:
        - installed
        - name: nginx
    service.running:
        - enable: true
        - name: nginx
        - require:
            - pkg: firewall
            - file: testfile
        - watch:
            - pkg: firewall
            - file: testfile

testfile:
  file.managed:
    - name: /tmp/file.test
    - source: salt://file.test
```

`shorewall.sls`
```
firewall:
    pkg:
        - installed
        - name: shorewall
    service.running:
        - enable: true
        - name: shorewall
        - require:
            - pkg: firewall
            - file: testfile
        - watch:
            - pkg: firewall
            - file: testfile

testfile:
  file.managed:
    - name: /tmp/file.test
    - source: salt://file.test
```

`file.test`
```
test1
```

nginx state output:
```
testrhodecode:
----------
          ID: firewall
    Function: pkg.installed
        Name: nginx
      Result: True
     Comment: Package nginx is already installed
     Changes:
----------
          ID: testfile
    Function: file.managed
        Name: /tmp/file.test
      Result: True
     Comment: File /tmp/file.test updated
     Changes:
              ----------
              diff:
                  ---
                  +++
                  @@ -1 +1 @@
                  -test14
                  +test1

----------
          ID: firewall
    Function: service.running
        Name: nginx
      Result: True
     Comment: Service restarted
     Changes:
              ----------
              nginx:
                  True

Summary
------------
Succeeded: 3
Failed:    0
------------
Total:     3
```

nginx minion log:
```
2014-11-20 09:42:08,611 [salt.minion                  ][INFO    ] User sudo_vlad Executing command state.sls with jid 20141120094208607503
2014-11-20 09:42:09,447 [salt.state                   ][INFO    ] Loading fresh modules for state activity
2014-11-20 09:42:09,815 [salt.fileclient              ][INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://nginx.sls'
2014-11-20 09:42:09,830 [salt.state                   ][INFO    ] Running state [nginx] at time 09:42:09.830383
2014-11-20 09:42:09,831 [salt.state                   ][INFO    ] Executing state pkg.installed for nginx
2014-11-20 09:42:09,832 [salt.loaded.int.module.cmdmod][INFO    ] Executing command "dpkg-query --showformat='${Status} ${Package} ${Version} ${Architecture}\n' -W" in directory '/root'
2014-11-20 09:42:09,861 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'grep-available -F Provides -s Package,Provides -e "^.+$"' in directory '/root'
2014-11-20 09:42:09,906 [salt.state                   ][INFO    ] Package nginx is already installed
2014-11-20 09:42:09,907 [salt.state                   ][INFO    ] Completed state [nginx] at time 09:42:09.907118
2014-11-20 09:42:09,909 [salt.state                   ][INFO    ] Running state [/tmp/file.test] at time 09:42:09.908716
2014-11-20 09:42:09,909 [salt.state                   ][INFO    ] Executing state file.managed for /tmp/file.test
2014-11-20 09:42:10,325 [salt.fileclient              ][INFO    ] Fetching file from saltenv 'base', ** done ** 'file.test'
2014-11-20 09:42:10,328 [salt.state                   ][INFO    ] File changed:
---
+++
@@ -1 +1 @@
-test14
+test1

2014-11-20 09:42:10,328 [salt.state                   ][INFO    ] Completed state [/tmp/file.test] at time 09:42:10.328397
2014-11-20 09:42:10,329 [salt.state                   ][INFO    ] Running state [nginx] at time 09:42:10.329215
2014-11-20 09:42:10,330 [salt.state                   ][INFO    ] Executing state service.running for nginx
2014-11-20 09:42:10,331 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'runlevel /run/utmp' in directory '/root'
2014-11-20 09:42:10,338 [salt.loaded.int.module.cmdmod][INFO    ] output: N 2
2014-11-20 09:42:10,351 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'service nginx status' in directory '/root'
2014-11-20 09:42:10,370 [salt.loaded.int.module.cmdmod][INFO    ] output:  * nginx is running
2014-11-20 09:42:10,377 [salt.state                   ][INFO    ] Service nginx is already enabled, and is in the desired state
2014-11-20 09:42:10,378 [salt.state                   ][INFO    ] Completed state [nginx] at time 09:42:10.377728
2014-11-20 09:42:10,378 [salt.state                   ][INFO    ] Running state [nginx] at time 09:42:10.378381
2014-11-20 09:42:10,379 [salt.state                   ][INFO    ] Executing state service.mod_watch for nginx
2014-11-20 09:42:10,382 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'service nginx status' in directory '/root'
2014-11-20 09:42:10,398 [salt.loaded.int.module.cmdmod][INFO    ] output:  * nginx is running
2014-11-20 09:42:10,399 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'service nginx restart' in directory '/root'
2014-11-20 09:42:11,465 [salt.loaded.int.module.cmdmod][INFO    ] output:  * Restarting nginx nginx
   ...done.
2014-11-20 09:42:11,466 [salt.state                   ][INFO    ] {'nginx': True}
2014-11-20 09:42:11,467 [salt.state                   ][INFO    ] Completed state [nginx] at time 09:42:11.466827
2014-11-20 09:42:11,469 [salt.minion                  ][INFO    ] Returning information for job: 20141120094208607503
```

shorewall state output:
```
testrhodecode:
----------
          ID: firewall
    Function: pkg.installed
        Name: shorewall
      Result: True
     Comment: Package shorewall is already installed
     Changes:
----------
          ID: testfile
    Function: file.managed
        Name: /tmp/file.test
      Result: True
     Comment: File /tmp/file.test updated
     Changes:
              ----------
              diff:
                  ---
                  +++
                  @@ -1 +1 @@
                  -test1
                  +test2

----------
          ID: firewall
    Function: service.running
        Name: shorewall
      Result: True
     Comment: Service shorewall has been enabled, and is in the desired state
     Changes:
              ----------
              shorewall:
                  True

Summary
------------
Succeeded: 3
Failed:    0
------------
Total:     3
```

shorewall minion log:
```
2014-11-20 09:46:51,445 [salt.minion                  ][INFO    ] User sudo_vlad Executing command state.sls with jid 20141120094651442368
2014-11-20 09:46:52,209 [salt.state                   ][INFO    ] Loading fresh modules for state activity
2014-11-20 09:46:52,559 [salt.fileclient              ][INFO    ] Fetching file from saltenv 'base', ** done ** 'shorewall.sls'
2014-11-20 09:46:52,572 [salt.state                   ][INFO    ] Running state [shorewall] at time 09:46:52.572026
2014-11-20 09:46:52,573 [salt.state                   ][INFO    ] Executing state pkg.installed for shorewall
2014-11-20 09:46:52,574 [salt.loaded.int.module.cmdmod][INFO    ] Executing command "dpkg-query --showformat='${Status} ${Package} ${Version} ${Architecture}\n' -W" in directory '/root'
2014-11-20 09:46:52,616 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'grep-available -F Provides -s Package,Provides -e "^.+$"' in directory '/root'
2014-11-20 09:46:52,660 [salt.state                   ][INFO    ] Package shorewall is already installed
2014-11-20 09:46:52,661 [salt.state                   ][INFO    ] Completed state [shorewall] at time 09:46:52.661291
2014-11-20 09:46:52,663 [salt.state                   ][INFO    ] Running state [/tmp/file.test] at time 09:46:52.662624
2014-11-20 09:46:52,663 [salt.state                   ][INFO    ] Executing state file.managed for /tmp/file.test
2014-11-20 09:46:53,071 [salt.fileclient              ][INFO    ] Fetching file from saltenv 'base', ** done ** 'file.test'
2014-11-20 09:46:53,074 [salt.state                   ][INFO    ] File changed:
---
+++
@@ -1 +1 @@
-test1
+test2

2014-11-20 09:46:53,075 [salt.state                   ][INFO    ] Completed state [/tmp/file.test] at time 09:46:53.075372
2014-11-20 09:46:53,076 [salt.state                   ][INFO    ] Running state [shorewall] at time 09:46:53.076127
2014-11-20 09:46:53,077 [salt.state                   ][INFO    ] Executing state service.running for shorewall
2014-11-20 09:46:53,079 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'runlevel /run/utmp' in directory '/root'
2014-11-20 09:46:53,086 [salt.loaded.int.module.cmdmod][INFO    ] output: N 2
2014-11-20 09:46:53,097 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'service shorewall status' in directory '/root'
2014-11-20 09:46:53,133 [salt.loaded.int.module.cmdmod][INFO    ] output: Shorewall-4.6.2.2 Status at testrhodecode - Thu Nov 20 09:46:53 CET 2014

Shorewall is running
State:Started (Thu Nov 20 09:40:44 CET 2014) from /etc/shorewall/ (/var/lib/shorewall/firewall compiled by Shorewall version 4.6.2.2)
2014-11-20 09:46:53,142 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'update-rc.d -f shorewall defaults' in directory '/root'
2014-11-20 09:46:53,192 [salt.loaded.int.module.cmdmod][INFO    ] output:
2014-11-20 09:46:53,193 [salt.state                   ][INFO    ] {'shorewall': True}
2014-11-20 09:46:53,194 [salt.state                   ][INFO    ] Completed state [shorewall] at time 09:46:53.193475
2014-11-20 09:46:53,196 [salt.minion                  ][INFO    ] Returning information for job: 20141120094651442368
```

shorewall output without `enable: true`:
```
testrhodecode:
----------
          ID: firewall
    Function: pkg.installed
        Name: shorewall
      Result: True
     Comment: Package shorewall is already installed
     Changes:
----------
          ID: testfile
    Function: file.managed
        Name: /tmp/file.test
      Result: True
     Comment: File /tmp/file.test updated
     Changes:
              ----------
              diff:
                  ---
                  +++
                  @@ -1 +1 @@
                  -test2
                  +test3

----------
          ID: firewall
    Function: service.running
        Name: shorewall
      Result: True
     Comment: Service restarted
     Changes:
              ----------
              shorewall:
                  True

Summary
------------
Succeeded: 3
Failed:    0
------------
Total:     3
```

shorewall log without `enable: true`:
```
2014-11-20 09:47:50,554 [salt.minion                  ][INFO    ] User sudo_vlad Executing command state.sls with jid 20141120094750551550
2014-11-20 09:47:51,377 [salt.state                   ][INFO    ] Loading fresh modules for state activity
2014-11-20 09:47:51,724 [salt.fileclient              ][INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://shorewall.sls'
2014-11-20 09:47:51,737 [salt.state                   ][INFO    ] Running state [shorewall] at time 09:47:51.737413
2014-11-20 09:47:51,738 [salt.state                   ][INFO    ] Executing state pkg.installed for shorewall
2014-11-20 09:47:51,741 [salt.loaded.int.module.cmdmod][INFO    ] Executing command "dpkg-query --showformat='${Status} ${Package} ${Version} ${Architecture}\n' -W" in directory '/root'
2014-11-20 09:47:51,769 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'grep-available -F Provides -s Package,Provides -e "^.+$"' in directory '/root'
2014-11-20 09:47:51,811 [salt.state                   ][INFO    ] Package shorewall is already installed
2014-11-20 09:47:51,813 [salt.state                   ][INFO    ] Completed state [shorewall] at time 09:47:51.812641
2014-11-20 09:47:51,814 [salt.state                   ][INFO    ] Running state [/tmp/file.test] at time 09:47:51.814192
2014-11-20 09:47:51,815 [salt.state                   ][INFO    ] Executing state file.managed for /tmp/file.test
2014-11-20 09:47:52,227 [salt.fileclient              ][INFO    ] Fetching file from saltenv 'base', ** done ** 'file.test'
2014-11-20 09:47:52,231 [salt.state                   ][INFO    ] File changed:
---
+++
@@ -1 +1 @@
-test2
+test3

2014-11-20 09:47:52,232 [salt.state                   ][INFO    ] Completed state [/tmp/file.test] at time 09:47:52.231820
2014-11-20 09:47:52,233 [salt.state                   ][INFO    ] Running state [shorewall] at time 09:47:52.232590
2014-11-20 09:47:52,233 [salt.state                   ][INFO    ] Executing state service.running for shorewall
2014-11-20 09:47:52,235 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'runlevel /run/utmp' in directory '/root'
2014-11-20 09:47:52,241 [salt.loaded.int.module.cmdmod][INFO    ] output: N 2
2014-11-20 09:47:52,251 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'service shorewall status' in directory '/root'
2014-11-20 09:47:52,287 [salt.loaded.int.module.cmdmod][INFO    ] output: Shorewall-4.6.2.2 Status at testrhodecode - Thu Nov 20 09:47:52 CET 2014

Shorewall is running
State:Started (Thu Nov 20 09:40:44 CET 2014) from /etc/shorewall/ (/var/lib/shorewall/firewall compiled by Shorewall version 4.6.2.2)
2014-11-20 09:47:52,288 [salt.state                   ][INFO    ] The service shorewall is already running
2014-11-20 09:47:52,289 [salt.state                   ][INFO    ] Completed state [shorewall] at time 09:47:52.288862
2014-11-20 09:47:52,290 [salt.state                   ][INFO    ] Running state [shorewall] at time 09:47:52.289885
2014-11-20 09:47:52,290 [salt.state                   ][INFO    ] Executing state service.mod_watch for shorewall
2014-11-20 09:47:52,292 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'service shorewall status' in directory '/root'
2014-11-20 09:47:52,327 [salt.loaded.int.module.cmdmod][INFO    ] output: Shorewall-4.6.2.2 Status at testrhodecode - Thu Nov 20 09:47:52 CET 2014

Shorewall is running
State:Started (Thu Nov 20 09:40:44 CET 2014) from /etc/shorewall/ (/var/lib/shorewall/firewall compiled by Shorewall version 4.6.2.2)
2014-11-20 09:47:52,329 [salt.loaded.int.module.cmdmod][INFO    ] Executing command 'service shorewall restart' in directory '/root'
2014-11-20 09:47:52,917 [salt.loaded.int.module.cmdmod][INFO    ] output: Restarting "Shorewall firewall": done.
2014-11-20 09:47:52,919 [salt.state                   ][INFO    ] {'shorewall': True}
2014-11-20 09:47:52,920 [salt.state                   ][INFO    ] Completed state [shorewall] at time 09:47:52.919547
2014-11-20 09:47:52,923 [salt.minion                  ][INFO    ] Returning information for job: 20141120094750551550
```

**Comment by roedie** [comment]
Just a simple 'me too' here.

I have exactly the same problem with Shorewall on Debian Jessie with Saltstack 2014.7.

If I remove the enable: True part everything works fine.

**Comment by roedie** [comment]
And then an update. If I switch from systemd to sysvinit-core everything works as expected.

Could it be it has something to do with: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=773392

**Comment by jfindlay** [comment]
Thank you all for your useful information on this issue.

**Comment by tramjoe** [comment]
Same problem here on Jessie + systemd, using salt 2014.7.15+ds-1. The service in question is Debian stock snmpd. The initscript behaves fine re. return codes, and has proper LSB default runlevels. I think this might be related to legacy script support in systemd.

**Comment by borutmrak** [comment]
I think this is the same problem that I'm having. I'm using the with postfix formula on Debian Jessie w/ systemd. If I use:

service.running:
  - enable: True

I get this on every run of this state:
[INFO    ] Executing command 'systemctl is-enabled postfix.service' in directory '/root'
[DEBUG   ] output: Failed to get unit file state for postfix.service: No such file or directory

So it seems systemd does not know when a legacy init script is enabled and returns false, salt re-enables it every time:

[INFO    ] Executing command 'systemctl enable postfix.service' in directory '/root'
[DEBUG   ] output: Synchronizing state for postfix.service with sysvinit using update-rc.d...
Executing /usr/sbin/update-rc.d postfix defaults
Executing /usr/sbin/update-rc.d postfix enable
[INFO    ] {'postfix': True}
[INFO    ] Completed state [postfix] at time 13:20:02.138909

So I think this is a systemd issue.

Any idea of a workaround?

**Comment by borutmrak** [comment]
Oh, and I've found this: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=705254

**Comment by blbradley** [comment]
Hi! I think I'm experiencing this bug on Debian Jessie as well. I'm using Debian packages from [Mesosphere](https://mesosphere.com/) which include ZooKeeper. They only support disabling services with `update-rc.d` because they do not support Wheezy yet. While trying apply `service.dead` with `enable=False` to the zookeeper service, the service is stopped but not disabled. When using the `debian_service` service provider, the opposite happened (service is disabled but not stopped).

Fortunately, you can separate this logic and use different service providers for each statement, like so:

```
stop-mesosphere-zookeeper-service:
  service.dead:
    - name: zookeeper


disable-mesosphere-zookeeper-service:
  service.disabled:
    - name: zookeeper
    - provider: debian_service
```

Cheers!

**Comment by bobrik** [comment]
Happened to me on Jessie as well. I install `ntp`, remove `/etc/init.d/ntp` and add `ntp.service` unit.

Here's what happens with `enable: true`:

```
salt λ cat ntp-wtf2.txt | fgrep systemctl
[INFO    ] Executing command 'systemctl --all --full --no-legend --no-pager list-units | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl --full --no-legend --no-pager list-unit-files | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl --all --full --no-legend --no-pager list-units | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl --full --no-legend --no-pager list-unit-files | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl status ntp.service' in directory '/root'
Warning: Unit file changed on disk, 'systemctl daemon-reload' recommended.
[INFO    ] Executing command 'systemctl --system daemon-reload' in directory '/root'
[INFO    ] Executing command 'systemctl is-active ntp.service' in directory '/root'
[INFO    ] Executing command 'systemctl --all --full --no-legend --no-pager list-units | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl --full --no-legend --no-pager list-unit-files | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl is-enabled ntp.service' in directory '/root'
[INFO    ] Executing command 'systemctl --all --full --no-legend --no-pager list-units | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl --full --no-legend --no-pager list-unit-files | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl status ntp.service' in directory '/root'
[INFO    ] Executing command 'systemctl enable ntp.service' in directory '/root'
[INFO    ] Executing command 'systemctl is-enabled ntp.service' in directory '/root'
```

And here with `enable: true` commented out:

```
[INFO    ] Executing command 'systemctl --all --full --no-legend --no-pager list-units | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl --full --no-legend --no-pager list-unit-files | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl --all --full --no-legend --no-pager list-units | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl --full --no-legend --no-pager list-unit-files | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl status ntp.service' in directory '/root'
Warning: Unit file changed on disk, 'systemctl daemon-reload' recommended.
[INFO    ] Executing command 'systemctl --system daemon-reload' in directory '/root'
[INFO    ] Executing command 'systemctl is-active ntp.service' in directory '/root'
[INFO    ] Executing command 'systemctl --all --full --no-legend --no-pager list-units | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl --full --no-legend --no-pager list-unit-files | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl status ntp.service' in directory '/root'
[INFO    ] Executing command 'systemctl is-active ntp.service' in directory '/root'
[INFO    ] Executing command 'systemctl --all --full --no-legend --no-pager list-units | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl --full --no-legend --no-pager list-unit-files | col -b' in directory '/root'
[INFO    ] Executing command 'systemctl status ntp.service' in directory '/root'
[INFO    ] Executing command 'systemctl restart ntp.service' in directory '/root'
```

Basically `enable: true` version never sees a restart:

```
          ID: ntp
    Function: service.running
      Result: True
     Comment: Service ntp has been enabled, and is in the desired state
     Started: 21:23:51.670309
    Duration: 518.623 ms
     Changes:
              ----------
              ntp:
                  True
```

I can provide full debug output for both scenarios, but I don't think it would help more than info provided already.

```
# salt --versions-report
           Salt: 2015.5.5
         Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
         Jinja2: 2.8
       M2Crypto: 0.21.1
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.7.0
           RAET: Not Installed
            ZMQ: 4.0.5
           Mako: Not Installed
        Tornado: Not Installed
        timelib: Not Installed
       dateutil: Not Installed
```

**Comment by bobrik** [comment]
Ha, look me, fixing this almost a year later: #34927.

**Comment by rallytime** [comment]
@bobrik Ah, thanks for finding this issue. For others on this thread, the PR referenced in the issue above is #35024. 

@vladvasiliu Does the fix in #35024 resolve this issue for you?

**Comment by Reiner030** [comment]
Since here / in referenced issue in last I cannot find the command here my tip. 

I got also in trouble of not starting daemons. They were in status "failed" and declined to start till I get after deeper searching with google the manual trick to run  
```systemctl reset-failed <service>```

which reset the failed state so they can started again normal.
